### PR TITLE
Add schema validation for our settings file

### LINKED ISSUE #2419: Error in settings results in strack trace distro import
### Describe the bug

Changing the default config from this:
```
mgmt_classes: []
mgmt_parameters:
 from_cobbler: 1
```

to this:
```
mgmt_classes: []
mgmt_parameters:
from_cobbler: 1
```

… results in this stack trace:

```
Wed Sep 16 17:15:17 2020 - INFO | Exception occured: <class 'AttributeError'>
Wed Sep 16 17:15:17 2020 - INFO | Exception value: 'NoneType' object has no attribute 'update'
Wed Sep 16 17:15:17 2020 - INFO | Exception Info:
  File "/usr/lib/python3.6/site-packages/cobbler/remote.py", line 2241, in _dispatch
    return method_handle(*params)

  File "/usr/lib/python3.6/site-packages/cobbler/remote.py", line 1063, in save_distro
    return self.save_item("distro", object_id, token, editmode=editmode)

  File "/usr/lib/python3.6/site-packages/cobbler/remote.py", line 1059, in save_item
    self.api.add_item(what, obj, logger=self.logger)

  File "/usr/lib/python3.6/site-packages/cobbler/api.py", line 437, in add_item
    self.get_items(what).add(ref, check_for_duplicate_names=check_for_duplicate_names, save=save, logger=logger)

  File "/usr/lib/python3.6/site-packages/cobbler/cobbler_collections/collection.py", line 358, in add
    self.lite_sync.add_single_distro(ref.name)

  File "/usr/lib/python3.6/site-packages/cobbler/actions/litesync.py", line 66, in add_single_distro
    self.tftpd.add_single_distro(distro)

  File "/usr/lib/python3.6/site-packages/cobbler/modules/managers/in_tftpd.py", line 137, in add_single_distro
    self.write_boot_files_distro(distro)

  File "/usr/lib/python3.6/site-packages/cobbler/modules/managers/in_tftpd.py", line 69, in write_boot_files_distro
    target = utils.blender(self.collection_mgr.api, False, distro)

  File "/usr/lib/python3.6/site-packages/cobbler/utils.py", line 580, in blender
    mgmt_parameters.update(results.get("autoinstall_meta", {}))
```

### To Reproduce

1. Change the config to **not** have the required indentation.
2. Import distro: `cobbler import --path /mnt --name test_distro`
3. Wait for error message.

### Expected behavior

Cobbler should instead complain about the settings.

### Version

```
Cobbler 3.1.2
```

### Additional context

SUSE employees can check [this bugzilla issue](https://bugzilla.suse.com/show_bug.cgi?id=1176295).

### PR DESCRIPTION
**Pull request by SchoolGuy**
This PR shall close #2419 

This PR also tries to pull all application settings related functionality inside the module and remove the factor that we require the file to be exactly at `/etc/cobbler/settings`. I see this as a first step into the direction of of containerized Cobbler.

### CODE REVIEW

**Comment by SchoolGuy** [comment]
Oh and I will try to sneak in some mongodb fixes. They are not important but since I am over that file anyway I won't ignore them. I can't break it more then it is already broken.

**Comment by SchoolGuy** [comment]
@Conan-Kudo Do we have any way to get https://github.com/keleshev/schema into all required RH distros? I see https://src.fedoraproject.org/rpms/python-schema but apparently this is not everywhere available for RH? openSUSE and Debian an good as far as I can see.

**Comment by SchoolGuy** [comment]
@dansou901 Can you make something out of this? - https://packages.debian.org/buster/python3-schema and https://github.com/cobbler/cobbler/pull/2530/checks?check_run_id=1808074146#step:3:6141

**Comment by dansou901** [comment]
Seems like python3-schema is missing from the apt install step in the Dockerfile. Add that and it should work.

**Comment by SchoolGuy** [comment]
Nice we have a circular dependency in `utils.py` and `settings.py`. This due to our unused multi-language support. I will remove that completely and will open a Feature Request for that so we can add it when it is useful again.

**Comment by SchoolGuy** [comment]
Codacy fails, yes this is bad, no I won't do more then I did because this PR is too huge anyway. I will split it later but that still won't fix Codacy. Thus we need to live with the fact that we aren't at a point where linters are useful. We will let them stay enabled because they are a good reminder where we can and need to improve.

**Comment by SchoolGuy** [comment]
I contacted the maintainer of `python-schema` and asked if it is possible to include this package in EPEL.

**Comment by SchoolGuy** [comment]
So python-schema has been added to EPEL 8. It is currently in testing but when it is in EPEL 8 fully, CentOS 8 should build also!

**Comment by SchoolGuy** [comment]
I don't know how to document this but we have to keep in mind that PyYAML can't write comments and can't read them. Thus our documentation in the YAML file will be gone once we write back to config once.

Source: https://github.com/yaml/pyyaml/issues/90

**Comment by SchoolGuy** [comment]
Rebase onto master and resolve conflict

**Comment by SchoolGuy** [comment]
Things that need to be done:

- [ ] Check the documentation and create hints so a manual migration of the settings file works
- [ ] Test: Add case so an old settings file is migrated correctly (also from `/etc/cobbler/settings`)
- [ ] Test: Unskip writing the settings file (thus replace it with the default one after the testcases)
- [ ] Code: Actually use the validation and handle errors if the settings file is invalid
- [ ] Test: Write tests with invalid settings files (especially with old style ones)

**Comment by SchoolGuy** [comment]
While reviewing the docs and related code I discovered that the settings docs are in a very poor state. I will address that along with this PR since it is a rework of our settings module.

**Review by brejoc — APPROVED** [review]
Pretty huge PR, but I haven't seen anything that led me to believe that there is a problem.

**Comment by SchoolGuy** [comment]
Rebase onto master

**Comment by SchoolGuy** [comment]
So old style settings are correctly migrated when they had default values before. I guess we need to bite in the apple that very custom settings may produce errors. This means that we need publish hotfix releases and prepare patches for distro maintainers in case anything goes wrong for users.

**Comment by SchoolGuy** [comment]
CentOS 8 again is making problems since a dependency of `python-schema` is now missing. I will try to fix that with some mails but this is definitely lowering my personal desire to be compatible with CentOS 8 RPMs...

**Comment by SchoolGuy** [comment]
@opoplawski Would you please be so kind to pull in the dependencies missing for `python-schema` into EPEL? CentOS 8 is kind of new territory for me, your name may open more doors then mine.

**Comment by SchoolGuy** [comment]
Cleaned Git History

File: config/cobbler/settings.d/manage_genders.settings [hunk 1]
```diff
@@ -1,2 +1,2 @@
 # manage_genders - Bool to enable/disable managing an /etc/genders file for use with pdsh and others.
-manage_genders: 0
+manage_genders: False
```
Comment (nodeg) [inline config/cobbler/settings.d/manage_genders.settings:2]:
```suggestion
manage_genders: false
```

File: docs/cobbler-conf.rst [hunk 2]
```diff
@@ -2,163 +2,288 @@
...
+  `INI <https://docs.python.org/3/library/configparser.html#supported-ini-file-structure>`_ specification.
+
+.. note:: Since we are cleaning a lot of tech-debt this may change over time. We are trying to find the balance which
+          format is the best for us to handle in the code and the best for admins to handle in the config files.
+
+.. warning:: If you are using ``allow_dynamic_settings``, then the comments in the YAML file will vanish after the first
+             change due to the fact that PyYAML doesn't support comments
+             (`Source <https://github.com/yaml/pyyaml/issues/90>`_)
+
+There are additional configuration files location which need to follow the YAML Syntax. These are loaded from the
```
Comment (nodeg) [inline docs/cobbler-conf.rst:18]:
```suggestion
There are additional configuration file locations which need to follow the YAML Syntax. These are loaded from the
```

File: docs/cobbler-conf.rst [hunk 3]
```diff
@@ -2,163 +2,288 @@
...
+
+Updates to the yaml-settings-file
+#################################
+
+Starting with 3.2.1:
+
+- We require the extension ``.yaml`` on our settings file to indicate the format of the file to editors and comply to
+  standards of the YAML specification.
+- We require the usage of Booleans in the format of ``True`` and ``False``. If you have old integer style booleans with
+  ``1`` and ``0`` this is fine but may should convert them as soon as possible. We may decide in a future version to
```
Comment (nodeg) [inline docs/cobbler-conf.rst:34]:
```suggestion
- We require the usage of booleans in the format of ``True`` and ``False``. If you have old integer style booleans with
  ``1`` and ``0`` this is fine but you may should convert them as soon as possible. We may decide in a future version to
```

File: docs/cobbler-conf.rst [hunk 4]
```diff
@@ -2,163 +2,288 @@
...
+  enforce our new way in a stricter manner. Automatic conversion is only done on a best-effort/available-resources
+  basis.
+- We enforce the types of values to the keys. Additional unexpected keys will throw errors. If you have those used in
+  Cobbler please report this in our Issue Tracker. We have decided to go this way to be able to rely on the existence
```
Comment (nodeg) [inline docs/cobbler-conf.rst:38]:
```suggestion
  Cobbler please report this in our issue tracker. We have decided to go this way to be able to rely on the existence
```

File: docs/cobbler-conf.rst [hunk 5]
```diff
@@ -2,163 +2,288 @@
...
+
+default: ``False``
 
 allow_duplicate_ips
 ===================
-if 1, Cobbler will allow insertions of system records that duplicate the IP address information of other system records.
-In general, this is undesirable and should be left 0.
 
-default: ``0``
+If "True", Cobbler will allow insertions of system records that duplicate the IP address information of other system
```
Comment (nodeg) [inline docs/cobbler-conf.rst:55]:
```suggestion
If ``True``, Cobbler will allow insertions of system records that duplicate the IP address information of other system
```

File: docs/cobbler-conf.rst [hunk 6]
```diff
@@ -2,163 +2,288 @@
...
+===============
+
+The location where Cobbler searches for the bootloaders to copy into the web directory.
+
+default: ``/var/lib/cobbler/loaders``
+
+grubconfig_dir
+==============
+
+The location where Cobbler searches for Grub configuration files.
```
Comment (nodeg) [inline docs/cobbler-conf.rst:158]:
```suggestion
The location where Cobbler searches for GRUB configuration files.
```

File: docs/cobbler-conf.rst [hunk 7]
```diff
@@ -2,163 +2,288 @@
...
+
+default: ``/var/lib/cobbler/grub_config``
+
 build_reporting_*
 =================
+
 Email out a report when Cobbler finishes installing a system.
 
-- enabled: set to 1 to turn this feature on
+- enabled: set to ``True`` to turn this feature on
```
Comment (nodeg) [inline docs/cobbler-conf.rst:167]:
```suggestion
- enabled: Set to ``True`` to turn this feature on
```

File: docs/cobbler-conf.rst [hunk 8]
```diff
@@ -2,163 +2,288 @@
...
+- email: Which addresses to email
+- ignorelist: TODO
 - sender: optional
```
Comment (nodeg) [inline docs/cobbler-conf.rst:170]:
```suggestion
- sender: Optional
```

File: docs/cobbler-conf.rst [hunk 9]
```diff
@@ -2,163 +2,288 @@
...
+- smtp_server: Used to specify another server for an MTA.
+- subject: Use the default subject unless overridden.
 
 defaults:
 
-.. code::
+.. code:: YAML
 
-    build_reporting_enabled: 0
+    build_reporting_enabled: False
```
Comment (nodeg) [inline docs/cobbler-conf.rst:178]:
```suggestion
    build_reporting_enabled: false
```

File: docs/cobbler-conf.rst [hunk 10]
```diff
@@ -2,163 +2,288 @@
...
     build_reporting_email: [ 'root@localhost' ]
     build_reporting_smtp_server: "localhost"
     build_reporting_subject: ""
     build_reporting_ignorelist: [ "" ]
 
+buildisodir
+===========
+
+Used for caching the intermediate files for ISO-Building. You may want to use an SSD, a tmpfs or something which is not
+persisted across reboots and can be easily thrown away but is also fast.
```
Comment (nodeg) [inline docs/cobbler-conf.rst:189]:
```suggestion
Used for caching the intermediate files for ISO-Building. You may want to use a SSD, a tmpfs or something which does not
persist across reboots and can be easily thrown away but is also fast.
```

File: docs/cobbler-conf.rst [hunk 11]
```diff
@@ -2,163 +2,288 @@
...
+If set to ``True``, all commands to the API (not directly to the XML-RPC server) will go over HTTPS instead of plain
+text. Be sure to change the ``http_port`` setting to the correct value for the web server.
+
+default: ``False``
+
+client_use_localhost
+====================
+
+If set to ``True``, all commands will be forced to use the localhost address instead of using the above value which can
+force commands like Cobbler sync to open a connection to a remote address if one is in the configuration and would
```
Comment (nodeg) [inline docs/cobbler-conf.rst:229]:
```suggestion
force commands like ``cobbler sync`` to open a connection to a remote address if one is in the configuration and would
```

File: docs/cobbler-conf.rst [hunk 12]
```diff
@@ -2,163 +2,288 @@
...
+==============
+
+Used for replicating the Cobbler instance.
+
+default: ``""``
+
+convert_server_to_ip
+====================
+
+Convert hostnames to IP-addresses (where possible) so DNS isn't a requirement for various tasks to work correctly.
```
Comment (nodeg) [inline docs/cobbler-conf.rst:244]:
```suggestion
Convert hostnames to IP addresses (where possible) so DNS isn't a requirement for various tasks to work correctly.
```

File: docs/cobbler-conf.rst [hunk 13]
```diff
@@ -168,69 +293,118 @@ default: ``"$1$mF86/UHC$WvcIcX2t6crBz2onWxyac."``
...
 =====================
+
 The default template type to use in the absence of any other detected template. If you do not specify the template
 with ``#template=<template_type>`` on the first line of your templates/snippets, Cobbler will assume try to use the
 following template engine to parse the templates.
 
-Current valid values are: cheetah, jinja2
+.. note:: Over time we will try to deprecate and remove Cheetah3 as a template engine. It is hard to package and there
+          are fewer guides then with Jinja2. Making the templating independent of the engine is a task which complicates
+          the code. Thus please try to use Jinja2. We will try to support a seamless transition on a best-effort basis.
```
Comment (nodeg) [inline docs/cobbler-conf.rst:303]:
```suggestion
          the code. Thus, please try to use Jinja2. We will try to support a seamless transition on a best-effort basis.
```

File: docs/cobbler-conf.rst [hunk 14]
```diff
@@ -248,147 +422,141 @@ defaults:
     ldap_server: "ldap.example.com"
     ldap_base_dn: "DC=example,DC=com"
     ldap_port: 389
-    ldap_tls: 1
-    ldap_anonymous_bind: 1
+    ldap_tls: True
+    ldap_anonymous_bind: True
```
Comment (nodeg) [inline docs/cobbler-conf.rst:426]:
```suggestion
    ldap_tls: true
    ldap_anonymous_bind: true
```

File: docs/cobbler-conf.rst [hunk 15]
```diff
@@ -248,147 +422,141 @@ defaults:
...
 
-mgmt_*
-======
-Cobbler has a feature that allows for integration with config management systems such as Puppet. The following
-parameters work in conjunction with ``--mgmt-classes`` and are described in further detail at :ref:`configuration-management`.
+bind_manage_ipmi
+================
 
-.. code-block:: Yaml
+When using the Bind9 DNS Server, you can enable or disable if the BMCs should receive own DNS entries.
```
Comment (nodeg) [inline docs/cobbler-conf.rst:437]:
```suggestion
When using the Bind9 DNS server, you can enable or disable if the BMCs should receive own DNS entries.
```

File: docs/cobbler-conf.rst [hunk 16]
```diff
@@ -248,147 +422,141 @@ defaults:
...
+    manage_forward_zones: []
+    manage_reverse_zones: []
+
+manage_genders
 ==============
-Let Cobbler know that you're using a newer version of puppet. Choose version 3 to use: 'puppet agent'; version 2 uses
-status quo: 'puppetd'. This one is commented out by default.
 
-default: ``2``
+Whether or not to manage the Genders file. For more information on that visit:
```
Comment (nodeg) [inline docs/cobbler-conf.rst:473]:
```suggestion
Whether or not to manage the genders file. For more information on that visit:
```

File: docs/cobbler-conf.rst [hunk 17]
```diff
@@ -248,147 +422,141 @@ defaults:
...
+Cobbler has a feature that allows for integration with config management systems such as Puppet. The following
+parameters work in conjunction with ``--mgmt-classes`` and are described in further detail at
+:ref:`configuration-management`.
 
-bind_master
+.. code-block:: Yaml
+
+    mgmt_classes: []
+    mgmt_parameters:
+        from_cobbler: True
```
Comment (nodeg) [inline docs/cobbler-conf.rst:504]:
```suggestion
.. code-block:: YAML

    mgmt_classes: []
    mgmt_parameters:
        from_cobbler: true
```

File: docs/cobbler-conf.rst [hunk 18]
```diff
@@ -248,147 +422,141 @@ defaults:
...
 
-manage_tftpd
-==============
-Set to 1 to enable Cobbler's TFTP management features. the choice of TFTP management engine is in
-``/etc/cobbler/modules.conf``.
+nsupdate_enabled
+================
 
-default: ``1``
+This enabled or disables the replacement (or removal) of records in the DNS zone for systems created (or removed) by
```
Comment (nodeg) [inline docs/cobbler-conf.rst:517]:
```suggestion
This enables or disables the replacement (or removal) of records in the DNS zone for systems created (or removed) by
```

File: docs/cobbler-conf.rst [hunk 19]
```diff
@@ -406,42 +574,105 @@ Choices (refer to codes.py):
...
 want this behavior and do not have a cross-multi-org separation concern. If you have a single org in your satellite,
 it's probably safe to turn this on and then you can use CobblerWeb alongside a Satellite install.
 
-default: ``0``
+default: ``False``
+
+redhat_management_server
+========================
+
+This setting is only used by the code that supports using Spacewalk/Satellite authentication within Cobbler Web and
```
Comment (nodeg) [inline docs/cobbler-conf.rst:668]:
```suggestion
This setting is only used by the code that supports using Uyuni/SUSE Manager/Spacewalk/Satellite authentication within Cobbler Web and
```

File: docs/cobbler-conf.rst [hunk 20]
```diff
@@ -504,50 +765,78 @@ default:
 
 .. code::
```
Comment (nodeg) [inline docs/cobbler-conf.rst:766]:
```suggestion
.. code:: YAML
```

File: docs/cobbler-conf.rst [hunk 21]
```diff
@@ -504,50 +765,78 @@ default:
...
 
-    scm_track_enabled: 0
+    scm_track_enabled: False
```
Comment (nodeg) [inline docs/cobbler-conf.rst:768]:
```suggestion
    scm_track_enabled: false
```

File: docs/cobbler-conf.rst [hunk 22]
```diff
@@ -482,20 +741,22 @@ defaults:
 
 .. code::
```
Comment (nodeg) [inline docs/cobbler-conf.rst:742]:
```suggestion
.. code:: YAML
```

File: docs/cobbler-conf.rst [hunk 23]
```diff
@@ -482,20 +741,22 @@ defaults:
...
 
-    restart_dns: 1
-    restart_dhcp: 1
+    restart_dns: True
+    restart_dhcp: True
```
Comment (nodeg) [inline docs/cobbler-conf.rst:745]:
```suggestion
    restart_dns: true
    restart_dhcp: true
```

File: docs/cobbler-conf.rst [hunk 24]
```diff
@@ -482,20 +741,22 @@ defaults:
...
 run_install_triggers
 ====================
+
 Install triggers are scripts in ``/var/lib/cobbler/triggers/install`` that are triggered in autoinstall pre and post
 sections. Any executable script in those directories is run. They can be used to send email or perform other actions.
 They are currently run as root so if you do not need this functionality you can disable it, though this will also
 disable ``cobbler status`` which uses a logging trigger to audit install progress.
 
-default: ``1``
+default: ``True``
```
Comment (nodeg) [inline docs/cobbler-conf.rst:755]:
```suggestion
default: ``true``
```

File: docs/cobbler-conf.rst [hunk 25]
```diff
@@ -504,50 +765,78 @@ default:
...
-default: ``0``
+When puppet starts on a system after installation it needs to have its certificate signed by the puppet master server.
+Enabling the following feature will ensure that the puppet server signs the certificate after installation if the puppet
+master server is running on the same machine as Cobbler. This requires ``puppet_auto_setup`` above to be enabled.
 
-client_use_https
-================
-If set to 1, all commands to the API (not directly to the XML-RPC server) will go over HTTPS instead of plain text. Be
-sure to change the ``http_port`` setting to the correct value for the web server.
+default: ``False``
```
Comment (nodeg) [inline docs/cobbler-conf.rst:797]:
```suggestion
default: ``false``
```

File: docs/cobbler-conf.rst [hunk 26]
```diff
@@ -504,50 +765,78 @@ default:
...
+
+signature_path
+==============
 
-default: ``0``
+The Cobbler Import workflow is powered by this file. Its location can be set with this config option.
```
Comment (nodeg) [inline docs/cobbler-conf.rst:802]:
```suggestion
The Cobbler import workflow is powered by this file. Its location can be set with this config option.
```

File: docs/cobbler-conf.rst [hunk 27]
```diff
@@ -504,50 +765,78 @@ default:
...
+Default: ``/srv/tftpboot``
 
 virt_auto_boot
 ==============
+
 Should new profiles for virtual machines default to auto booting with the physical host when the physical host reboots?
 This can be overridden on each profile or system object.
 
-default: ``1``
+default: ``True``
```
Comment (nodeg) [inline docs/cobbler-conf.rst:827]:
```suggestion
default: ``true``
```

File: docs/cobbler-conf.rst [hunk 28]
```diff
@@ -572,99 +861,48 @@ default:
...
-Flags to use for yumdownloader. Not all versions may support ``--resolve``.
-
-default: ``"--resolve"``
-
-serializer_pretty_json
-======================
-Sort and indent JSON output to make it more human-readable.
-
-default: ``0``
+default: ``True``
```
Comment (nodeg) [inline docs/cobbler-conf.rst:876]:
```suggestion
default: ``true``
```

File: docs/cobbler-conf.rst [hunk 29]
```diff
@@ -572,99 +861,48 @@ default:
...
-=============
-Internal proxy - used by systems to reach Cobbler for kickstarts.
-
-E.g.: proxy_url_int: ``http://10.0.0.1:8080``
+``cobbler repo add`` commands set Cobbler up with repository information that can be used during autoinstall and is
+automatically set up in the Cobbler autoinstall templates. By default, these are only available at install time. To
+make these repositories usable on installed systems (since Cobbler makes a very convenient mirror) set this to ``True``.
+Most users can safely set this to ``True``. Users who have a dual homed Cobbler server, or are installing laptops that
+will not always have access to the Cobbler server may wish to leave this as ``False``. In that case, the Cobbler
+mirrored yum repos are still accessible at ``http://cobbler.example.org/cblr/repo_mirror`` and yum configuration can
```
Comment (nodeg) [inline docs/cobbler-conf.rst:886]:
```suggestion
mirrored yum repos are still accessible at ``http://cobbler.example.org/cblr/repo_mirror`` and YUM configuration can
```

File: docs/cobbler-conf.rst [hunk 30]
```diff
@@ -747,7 +989,9 @@ default: ``manage_isc``
 
 tftpd
 =====
-Chooses the TFTP management engine if manage_tftp is enabled in ``/etc/cobbler/settings.yaml``, which is ON by default.
+
+Chooses the TFTP management engine if ``manage_tftpd`` is enabled in ``/etc/cobbler/settings.yaml``, which is ON by
```
Comment (nodeg) [inline docs/cobbler-conf.rst:993]:
```suggestion
Chooses the TFTP management engine if ``manage_tftpd`` is enabled in ``/etc/cobbler/settings.yaml``, which is **on** by
```

**Review by nodeg — CHANGES_REQUESTED** [review]
I left some comments.

**Comment by opoplawski** [comment]
> CentOS 8 again is making problems since a dependency of `python-schema` is now missing. I will try to fix that with some mails but this is definitely lowering my personal desire to be compatible with CentOS 8 RPMs...

I'm committed to making cobbler work on CentOS 8.  I've piled on to request for the needed dep to get added and offered to help out.

**Comment by SchoolGuy** [comment]
@opoplawski Thanks a lot! That is very great news for me! :)

File: cobbler/settings.py [hunk 31]
```diff
@@ -367,24 +528,53 @@ def read_settings_file(filepath="/etc/cobbler/settings.yaml") -> Union[Dict[Hash
...
-    except yaml.YAMLError as e:
+    except yaml.YAMLError as error:
         traceback.print_exc()
-        raise CX("\"%s\" is not a valid YAML file" % filepath) from e
+        raise yaml.YAMLError("\"%s\" is not a valid YAML file" % filepath) from error
+    filecontent = __migrate_settingsfile_int_bools(filecontent)
+    try:
+        validate_settings(filecontent)
+    except SchemaMissingKeyError:
+        logging.exception("Settingsfile was not returned due to missing Keys.")
```
Comment (nodeg) [inline cobbler/settings.py:538]:
```suggestion
        logging.exception("Settings file was not returned due to missing Keys.")
```

File: cobbler/settings.py [hunk 32]
```diff
@@ -367,24 +528,53 @@ def read_settings_file(filepath="/etc/cobbler/settings.yaml") -> Union[Dict[Hash
...
+        logging.debug("The settings to read were: \"%s\"", filecontent)
+        return {}
+    except SchemaError:
+        logging.exception("Settingsfile was returned due to an error in the schema.")
```
Comment (nodeg) [inline cobbler/settings.py:542]:
```suggestion
        logging.exception("Settings file was returned due to an error in the schema.")
```

File: cobbler/settings.py [hunk 33]
```diff
@@ -367,24 +528,53 @@ def read_settings_file(filepath="/etc/cobbler/settings.yaml") -> Union[Dict[Hash
...
+        with open(filepath, "w") as settings_file:
+            yaml_dump = yaml.safe_dump(validated_data)
+            header = "# Cobbler settings file\n"
+            header += "# Docs for this file can be found at: https://cobbler.readthedocs.io/en/latest/cobbler-conf.html"
+            header += "\n\n"
+            yaml_dump = header + yaml_dump
+            settings_file.write(yaml_dump)
+        return True
+    except SchemaMissingKeyError:
+        logging.exception("Settingsfile was not written to the disc due to missing Keys.")
```
Comment (nodeg) [inline cobbler/settings.py:570]:
```suggestion
        logging.exception("Settings file was not written to the disc due to missing keys.")
```

File: cobbler/settings.py [hunk 34]
```diff
@@ -367,24 +528,53 @@ def read_settings_file(filepath="/etc/cobbler/settings.yaml") -> Union[Dict[Hash
...
+        logging.debug("The settings to write were: \"%s\"", data)
+        return False
+    except SchemaError:
+        logging.exception("Settingsfile was not written to the disc due to an error in the schema.")
```
Comment (nodeg) [inline cobbler/settings.py:574]:
```suggestion
        logging.exception("Settings file was not written to the disc due to an error in the schema.")
```

**Review by nodeg — APPROVED** [review]

**Comment by opoplawski** [comment]
contextlib2 is now in EPEL testing - please provide karma here: https://bodhi.fedoraproject.org/updates/FEDORA-EPEL-2021-65e082736e

**Pull request merged**